pkg/agent/controller/networkpolicy/node_reconciler_linux.go

```diff
@@ -0,0 +1,703 @@
+		}
+		// Iterate every existing rules. If an existing rule is in the iptRulesToUpdate map, replace it with the new rule.
+		for index, iptRule := range cachedIPTRules {
+			if _, exists := iptRulesToUpdate[iptRule.ruleID]; exists {
+				cachedIPTRules[index] = iptRulesToUpdate[iptRule.ruleID]
+			}
+		}
+	} else {
+		// If these are new rules, append the new rules then sort all rules.
+		cachedIPTRules = append(cachedIPTRules, iptRules...)
+		sort.Slice(cachedIPTRules, func(i, j int) bool {
+			return !cachedIPTRules[i].priority.Less(*cachedIPTRules[j].priority)
+		})
+	}
+
+	// Get all the sorted iptables rules and synchronize them.
+	var iptRuleStrs []string
+	for _, r := range cachedIPTRules {
+		iptRuleStrs = append(iptRuleStrs, r.ruleStr)
+	}
+	if err := r.routeClient.AddOrUpdateNodeNetworkPolicyIPTables([]string{iptChain}, [][]string{iptRuleStrs}, isIPv6); err != nil {
+		return err
+	}
+
+	// cache the new iptables rules.
+	cachedCoreIPTChain.rules = cachedIPTRules
+	return nil
+}
+
+func (r *nodeReconciler) deleteCoreIPRule(ruleID string, iptChain string, isIPv6 bool) error {
+	cachedCoreIPTChain := r.getCachedCoreIPTChain(iptChain, isIPv6)
+	cachedCoreIPTChain.Lock()
+	defer cachedCoreIPTChain.Unlock()
+
+	// Get all cached iptables rules, then delete the rule with the given ruleID.
+	cachedIPTRules := cachedCoreIPTChain.rules
+	var indexToDelete int
+	for i := 0; i < len(cachedIPTRules); i++ {
+		if cachedIPTRules[i].ruleID == ruleID {
+			indexToDelete = i
+			break
+		}
+	}
+	cachedIPTRules = append(cachedIPTRules[:indexToDelete], cachedIPTRules[indexToDelete+1:]...)
+
+	// Get all the sorted iptables rules and synchronize them.
+	var iptRuleStrs []string
+	for _, r := range cachedIPTRules {
+		iptRuleStrs = append(iptRuleStrs, r.ruleStr)
+	}
+	if err := r.routeClient.AddOrUpdateNodeNetworkPolicyIPTables([]string{iptChain}, [][]string{iptRuleStrs}, isIPv6); err != nil {
+		return err
+	}
+
+	// cache the new iptables rules.
+	cachedCoreIPTChain.rules = cachedIPTRules
+	return nil
+}
+
+func (r *nodeReconciler) getCachedCoreIPTChain(iptChain string, isIPv6 bool) *coreIPTChain {
+	// There are 4 categories of cached core iptables rules:
+	// - For IPv4, iptables rules installed in chain ANTREA-INGRESS-RULES for ingress rules.
+	// - For IPv6, ip6tables rules installed in chain ANTREA-INGRESS-RULES for ingress rules.
+	// - For IPv4, iptables rules installed in chain ANTREA-EGRESS-RULES for egress rules.
+	// - For IPv6, ip6tables rules installed in chain ANTREA-EGRESS-RULES for egress rules.
+	categoryKey := genCacheCategory(iptChain, isIPv6)
+	return r.cachedCoreIPTChains[categoryKey]
+}
+
+func groupMembersToIPNets(groups v1beta2.GroupMemberSet, isIPv6 bool) sets.Set[string] {
+	ipNets := sets.New[string]()
+	suffix := "/32"
+	if isIPv6 {
+		suffix = "/128"
+	}
+	for _, member := range groups {
+		for _, ip := range member.IPs {
+			ipAddr := net.IP(ip)
+			if isIPv6 == utilnet.IsIPv6(ipAddr) {
```

Comment:
Do you need to check this? or here we can use `isIPv6 ` directly?

Reply:
This is from `isIPv6 && utilnet.IsIPv6(ipAddr) || !isIPv6 == !utilnet.IsIPv6(ipAddr)`

Reply:
Yeah, I mean will the member IPs be none IPv6 address when isIPv6 is true? or it's possible to be a invalid address? Do we validate the value in the rule before reconciling?

Reply:
I doesn't find such validation in previous code of ANP or ACNP.

Reply:
Ok, then let's keep as it is.